components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java

```diff
@@ -835,6 +850,88 @@ private void checkNullIdentityProvider(Assertion assertion, String tenantDomain,
         }
     }
 
+    /**
+     * If the token issuer validation fails against the resident identity provider's IdP Entity ID, checks whether the
+     * token issuer has been overriden by an "Idp Entity ID Alias" specified in a SAML SSO configuration. The check
+     * is done against IdP Entity ID Alias values of SAML SSO configurations in the registry. The SAML SSO configurations
+     * that needs to be checked are identified from the audience restrictions in the SAML assertion as it contains the
+     * "issuer" of the SAML SSO configuration with respective to the SAML SP.
+     *
+     * @param assertion
+     * @param tenantDomain
+     * @return
+     * @throws IdentityOAuth2Exception
+     */
+    private Boolean validateIdpEntityIdAliasFromSAMLSP(Assertion assertion, String tenantDomain)
+            throws IdentityOAuth2Exception {
+
+        Conditions conditions = assertion.getConditions();
+        List<AudienceRestriction> audienceRestrictions = conditions.getAudienceRestrictions();
+        validateAudienceRestriction(audienceRestrictions);
+        for (AudienceRestriction audienceRestriction : audienceRestrictions) {
+            if (CollectionUtils.isNotEmpty(audienceRestriction.getAudiences())) {
+                for (Audience audience : audienceRestriction.getAudiences()) {
+                    SAMLSSOServiceProviderDO samlssoServiceProviderDO = getSAMLSSOServiceProvider
+                            (audience.getAudienceURI(), tenantDomain);
+                    if (samlssoServiceProviderDO != null) {
+                        if (samlssoServiceProviderDO.getIdpEntityIDAlias() != null &&
+                                samlssoServiceProviderDO.getIdpEntityIDAlias().equals(assertion.getIssuer().
+                                        getValue())) {
+                            if (log.isDebugEnabled()) {
+                                log.debug("Token Issuer verified against IdP Entity ID Alias : " +
+                                        samlssoServiceProviderDO.getIdpEntityIDAlias() + " of SAML Service Provider " +
+                                        samlssoServiceProviderDO.getIssuer() + " in tenant : " + tenantDomain + ".");
+                            }
+                            return true;
+                        }
+                    }
+                }
+            }
+        }
+        if (log.isDebugEnabled()) {
+            log.debug("No SAML Service Provider configuration with IdP Entity ID Alias " +
+                    "similar to token issuer found.");
+        }
+        return false;
+    }
+
+    private SAMLSSOServiceProviderDO getSAMLSSOServiceProvider(String issuerName, String tenantDomain)
+            throws IdentityOAuth2Exception {
```

Comment:
Add some method-level comments

Reply:
Addressed in https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1438